crates/router/src/core/customers.rs

```diff
@@ -613,143 +659,394 @@ pub async fn delete_customer(
+                        self.key_store,
+                    )
                     .await
                     .switch()?,
-            ),
-            None => None,
-        }
-    };
-    let encrypted_data = types::crypto_operation(
-        &(&state).into(),
-        type_name!(domain::Customer),
-        types::CryptoOperation::BatchEncrypt(CustomerRequestWithEmail::to_encryptable(
-            CustomerRequestWithEmail {
-                name: update_customer.name.clone(),
-                email: update_customer.email.clone(),
-                phone: update_customer.phone.clone(),
-            },
-        )),
-        Identifier::Merchant(key_store.merchant_id.clone()),
-        key,
-    )
-    .await
-    .and_then(|val| val.try_into_batchoperation())
-    .switch()?;
-    let encryptable_customer = CustomerRequestWithEmail::from_encryptable(encrypted_data)
-        .change_context(errors::CustomersErrorResponse::InternalServerError)?;
+                ),
+                None => None,
+            }
+        };
+        Ok(address)
+    }
+}
 
-    let response = db
-        .update_customer_by_customer_id_merchant_id(
+struct VerifyIdForUpdateCustomer<'a> {
+    merchant_reference_id: Option<&'a id_type::CustomerId>,
+    id: Option<&'a String>,
+    merchant_account: &'a domain::MerchantAccount,
+    key_store: &'a domain::MerchantKeyStore,
+    key_manager_state: &'a KeyManagerState,
+}
+
+#[cfg(all(any(feature = "v1", feature = "v2"), not(feature = "customer_v2")))]
+impl<'a> VerifyIdForUpdateCustomer<'a> {
+    async fn verify_id_and_get_customer_object(
+        &self,
+        db: &dyn StorageInterface,
+    ) -> Result<domain::Customer, error_stack::Report<errors::CustomersErrorResponse>> {
+        let customer_id = self
+            .merchant_reference_id
+            .get_required_value("customer_id")
+            .change_context(errors::CustomersErrorResponse::InternalServerError)
+            .attach("Missing required field `customer_id`")?;
+
+        let _id = self.id;
+
+        let customer = db
+            .find_customer_by_customer_id_merchant_id(
+                self.key_manager_state,
+                customer_id,
+                self.merchant_account.get_id(),
+                self.key_store,
+                self.merchant_account.storage_scheme,
+            )
+            .await
+            .switch()?;
+
+        Ok(customer)
+    }
+}
+
+#[cfg(all(feature = "v2", feature = "customer_v2"))]
+impl<'a> VerifyIdForUpdateCustomer<'a> {
+    async fn verify_id_and_get_customer_object(
+        &self,
+        db: &dyn StorageInterface,
+    ) -> Result<domain::Customer, error_stack::Report<errors::CustomersErrorResponse>> {
+        let id = self
+            .id
+            .get_required_value("id")
+            .change_context(errors::CustomersErrorResponse::InternalServerError)
+            .attach("Missing required field `id`")?;
+
+        let _merchant_reference_id = self.merchant_reference_id;
```

Comment:
We should add a validation that if the current object in the storage layer has merchant_reference_id then compare both, if not matching throw a error.

If the current object in the storage layer don't have associated merchant_reference_id(i.e None) then don't throw a error